**<h1>Analiza Ekstraklasy od sezonu 2012/2013 do 2024/2025</h1>**


## **Opis projektu:** ##
Projekt przestawia szczegółową analize wszystkich meczów z ostatnich 13 sezonów w najlepszej polskiej lidze piłki nożnej czyli Ekstraklasy.
Zbiór danych został pobrany z realnych danych ze strony: [Football data](https://www.football-data.co.uk/poland.php)

**Logo Ekstraklasy:**

<img src="https://www.tygodnikprzeglad.pl/wp-content/uploads/2021/11/ekstraklasa-typy-dnia-800x500_c.jpg" alt="Logo ekstraklasy" width="400">


### **Opis kolumn:** ###

**Country** – kraj (tu: zawsze "Poland")

**League** – nazwa ligi (np. "Ekstraklasa")

**Season** – sezon rozgrywkowy (np. "2012/2013")

**Date** – data meczu

**Time** – godzina rozpoczęcia meczu

**Drużyny i wynik:**

**Home** – drużyna gospodarzy

**Away** – drużyna gości

**HG** – liczba goli zdobytych przez gospodarzy (Home Goals)

**AG** – liczba goli zdobytych przez gości (Away Goals)

**Res** – wynik końcowy z perspektywy gospodarza:

**"H"** – wygrana gospodarzy (Home win)

**"D"** – remis (Draw)

**"A"** – wygrana gości (Away win)

**Kursy bukmacherskie (od firm bukmacherskich):**

**Pinnacle Sports (PS):**

**PSCH** – kurs na wygraną gospodarzy

**PSCD** – kurs na remis

**PSCA** – kurs na wygraną gości

**Najwyższe kursy z wielu źródeł:**

**MaxCH** – najwyższy kurs na wygraną gospodarzy

**MaxCD** – najwyższy kurs na remis

**MaxCA** – najwyższy kurs na wygraną gości

**Średnie kursy:**

**AvgCH** – średni kurs na wygraną gospodarzy

**AvgCD** – średni kurs na remis

**AvgCA** – średni kurs na wygraną gości

**Betfair Exchange (prawdopodobnie rynki giełdowe zakładów):**

**BFECH** – kurs Betfair na gospodarzy

**BFECD** – kurs Betfair na remis

**BFECA** – kurs Betfair na gości  
(Uwaga: dostępne tylko dla ok. 300 meczów – dużo wartości NaN)

In [144]:
import pandas as pd
import sys

In [145]:
df = pd.read_excel(r"C:\Users\huber\Downloads\POL.xlsx")
df

Country       League     Season       Date      Time            Home  \
0     Poland  Ekstraklasa  2012/2013 2012-08-17  17:00:00  Pogon Szczecin   
1     Poland  Ekstraklasa  2012/2013 2012-08-17  19:30:00   Piast Gliwice   
2     Poland  Ekstraklasa  2012/2013 2012-08-18  12:30:00     Jagiellonia   
3     Poland  Ekstraklasa  2012/2013 2012-08-18  14:45:00     Widzew Lodz   
4     Poland  Ekstraklasa  2012/2013 2012-08-18  17:00:00     Lech Poznan   
...      ...          ...        ...        ...       ...             ...   
3771  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00           Legia   
3772  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00         Puszcza   
3773  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00  Radomiak Radom   
3774  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00           Rakow   
3775  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00        Zaglebie   

               Away  HG  AG Res  PSCH  PSCD   PSCA  MaxCH  MaxCD  MaxCA  \
0          Zaglebie   4   0   H  3.23  3.21   2.47   3.30   3.30   2.50   
1         Gornik Z.   1   2   A  2.72  3.12   2.98   2.78   3.25   3.30   
2      Podbeskidzie   2   1   H  1.74  3.60   5.63   1.85   3.60   5.63   
3     Slask Wroclaw   2   1   H  2.70  3.11   2.95   2.80   3.20   2.95   
4              Ruch   4   0   H  2.12  3.29   3.89   2.12   3.30   4.00   
...             ...  ..  ..  ..   ...   ...    ...    ...    ...    ...   
3771    Stal Mielec   2   2   D  1.35  5.47   7.60   1.44   5.50   7.70   
3772  Slask Wroclaw   1   1   D  3.25  3.93   2.05   3.30   4.00   2.05   
3773   Motor Lublin   2   3   A  1.93  3.64   3.86   1.99   3.75   4.05   
3774    Widzew Lodz   2   1   H  1.26  5.64  11.81   1.29   5.80  11.81   
3775       Cracovia   1   2   A  2.77  3.45   2.51   2.77   3.55   2.56   

      AvgCH  AvgCD  AvgCA  BFECH  BFECD  BFECA  
0      3.01   3.11   2.34    NaN    NaN    NaN  
1      2.56   3.05   2.78    NaN    NaN    NaN  
2      1.74   3.42   4.72    NaN    NaN    NaN  
3      2.66   3.03   2.66    NaN    NaN    NaN  
4      2.05   3.16   3.61    NaN    NaN    NaN  
...     ...    ...    ...    ...    ...    ...  
3771   1.36   5.29   7.12   1.40   5.60   8.80  
3772   3.24   3.88   2.01   3.55   4.10   2.08  
3773   1.94   3.54   3.73   2.02   3.80   3.90  
3774   1.27   5.47  10.29   1.30   6.20  13.50  
3775   2.67   3.45   2.49   2.86   3.65   2.60  

[3776 rows x 22 columns]

In [146]:
pd.set_option('display.max_columns', 25)

In [147]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3776 entries, 0 to 3775
Data columns (total 22 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   Country  3776 non-null   object        
 1   League   3776 non-null   object        
 2   Season   3776 non-null   object        
 3   Date     3776 non-null   datetime64[ns]
 4   Time     3776 non-null   object        
 5   Home     3776 non-null   object        
 6   Away     3776 non-null   object        
 7   HG       3776 non-null   int64         
 8   AG       3776 non-null   int64         
 9   Res      3776 non-null   object        
 10  PSCH     3775 non-null   float64       
 11  PSCD     3775 non-null   float64       
 12  PSCA     3775 non-null   float64       
 13  MaxCH    3776 non-null   float64       
 14  MaxCD    3776 non-null   float64       
 15  MaxCA    3776 non-null   float64       
 16  AvgCH    3776 non-null   float64       
 17  AvgCD    3776 non-null   float64 

In [148]:
df.shape

(3776, 22)

In [149]:
df.head(6)

Country       League     Season       Date      Time            Home  \
0  Poland  Ekstraklasa  2012/2013 2012-08-17  17:00:00  Pogon Szczecin   
1  Poland  Ekstraklasa  2012/2013 2012-08-17  19:30:00   Piast Gliwice   
2  Poland  Ekstraklasa  2012/2013 2012-08-18  12:30:00     Jagiellonia   
3  Poland  Ekstraklasa  2012/2013 2012-08-18  14:45:00     Widzew Lodz   
4  Poland  Ekstraklasa  2012/2013 2012-08-18  17:00:00     Lech Poznan   
5  Poland  Ekstraklasa  2012/2013 2012-08-19  13:30:00           Wisla   

            Away  HG  AG Res  PSCH  PSCD  PSCA  MaxCH  MaxCD  MaxCA  AvgCH  \
0       Zaglebie   4   0   H  3.23  3.21  2.47   3.30   3.30   2.50   3.01   
1      Gornik Z.   1   2   A  2.72  3.12  2.98   2.78   3.25   3.30   2.56   
2   Podbeskidzie   2   1   H  1.74  3.60  5.63   1.85   3.60   5.63   1.74   
3  Slask Wroclaw   2   1   H  2.70  3.11  2.95   2.80   3.20   2.95   2.66   
4           Ruch   4   0   H  2.12  3.29  3.89   2.12   3.30   4.00   2.05   
5  GKS Belchatow   2   1   H  1.50  4.13  8.31   1.55   4.20   8.31   1.50   

   AvgCD  AvgCA  BFECH  BFECD  BFECA  
0   3.11   2.34    NaN    NaN    NaN  
1   3.05   2.78    NaN    NaN    NaN  
2   3.42   4.72    NaN    NaN    NaN  
3   3.03   2.66    NaN    NaN    NaN  
4   3.16   3.61    NaN    NaN    NaN  
5   3.88   6.48    NaN    NaN    NaN

In [150]:
df.tail(6)

Country       League     Season       Date      Time            Home  \
3770  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00   Lechia Gdansk   
3771  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00           Legia   
3772  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00         Puszcza   
3773  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00  Radomiak Radom   
3774  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00           Rakow   
3775  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00        Zaglebie   

               Away  HG  AG Res  PSCH  PSCD   PSCA  MaxCH  MaxCD  MaxCA  \
3770   GKS Katowice   2   3   A  1.97  3.98   3.42   2.01   4.10   3.55   
3771    Stal Mielec   2   2   D  1.35  5.47   7.60   1.44   5.50   7.70   
3772  Slask Wroclaw   1   1   D  3.25  3.93   2.05   3.30   4.00   2.05   
3773   Motor Lublin   2   3   A  1.93  3.64   3.86   1.99   3.75   4.05   
3774    Widzew Lodz   2   1   H  1.26  5.64  11.81   1.29   5.80  11.81   
3775       Cracovia   1   2   A  2.77  3.45   2.51   2.77   3.55   2.56   

      AvgCH  AvgCD  AvgCA  BFECH  BFECD  BFECA  
3770   1.93   3.91   3.43   2.04   4.10   3.55  
3771   1.36   5.29   7.12   1.40   5.60   8.80  
3772   3.24   3.88   2.01   3.55   4.10   2.08  
3773   1.94   3.54   3.73   2.02   3.80   3.90  
3774   1.27   5.47  10.29   1.30   6.20  13.50  
3775   2.67   3.45   2.49   2.86   3.65   2.60

**Filtr - gospodarz, gość, data**

In [151]:
df.filter(items=['Home', 'Away', 'Date'], axis=1)

Home           Away       Date
0     Pogon Szczecin       Zaglebie 2012-08-17
1      Piast Gliwice      Gornik Z. 2012-08-17
2        Jagiellonia   Podbeskidzie 2012-08-18
3        Widzew Lodz  Slask Wroclaw 2012-08-18
4        Lech Poznan           Ruch 2012-08-18
...              ...            ...        ...
3771           Legia    Stal Mielec 2025-05-24
3772         Puszcza  Slask Wroclaw 2025-05-24
3773  Radomiak Radom   Motor Lublin 2025-05-24
3774           Rakow    Widzew Lodz 2025-05-24
3775        Zaglebie       Cracovia 2025-05-24

[3776 rows x 3 columns]

In [152]:
first_season = df["Season"].iloc[0]
last_season = df["Season"].iloc[3775] # Opcjonalnie -1    
print("First season:", first_season) 
print("Last season:", last_season)

First season: 2012/2013
Last season: 2024/2025


In [153]:
middle_index = len(df) // 2
print("Middle index:", middle_index)
df.iloc[1888]

Middle index: 1888


Country                 Poland
League             Ekstraklasa
Season               2018/2019
Date       2019-02-15 00:00:00
Time                  17:00:00
Home            Pogon Szczecin
Away                 Gornik Z.
HG                           3
AG                           1
Res                          H
PSCH                      1.75
PSCD                      3.67
PSCA                      5.26
MaxCH                     1.75
MaxCD                      3.8
MaxCA                     5.26
AvgCH                     1.71
AvgCD                     3.62
AvgCA                     4.81
BFECH                      NaN
BFECD                      NaN
BFECA                      NaN
Name: 1888, dtype: object

**Zdarzenia z najniższym kursem na gospodarzy - najbardziej pewne mecze gdzie gospodarza ma wygrać**

In [154]:
df[df["PSCH"] <= 1.2]

Country       League     Season       Date      Time         Home  \
1334  Poland  Ekstraklasa  2016/2017 2017-03-19  14:30:00  Lech Poznan   
2782  Poland  Ekstraklasa  2021/2022 2022-03-19  19:00:00  Lech Poznan   
2813  Poland  Ekstraklasa  2021/2022 2022-04-20  17:00:00  Lech Poznan   
2820  Poland  Ekstraklasa  2021/2022 2022-04-24  11:30:00  Lech Poznan   
2962  Poland  Ekstraklasa  2022/2023 2022-10-10  17:00:00        Rakow   
3066  Poland  Ekstraklasa  2022/2023 2023-03-10  19:30:00        Rakow   

               Away  HG  AG Res  PSCH  PSCD   PSCA  MaxCH  MaxCD  MaxCA  \
1334         Leczna   0   0   D  1.20  7.21  17.87   1.20   7.62  20.00   
2782    Jagiellonia   3   0   H  1.20  7.22  15.11   1.24   7.30  16.00   
2813         Leczna   3   0   H  1.16  7.79  23.43   1.16   8.00  23.43   
2820    Stal Mielec   3   1   H  1.20  7.11  16.98   1.22   7.11  17.07   
2962        Legnica   1   0   H  1.16  7.87  18.67   1.23   7.87  18.67   
3066  Slask Wroclaw   4   1   H  1.19  6.90  16.98   1.27   7.30  19.50   

      AvgCH  AvgCD  AvgCA  BFECH  BFECD  BFECA  
1334   1.18   6.51  15.22    NaN    NaN    NaN  
2782   1.18   6.78  13.94    NaN    NaN    NaN  
2813   1.14   7.30  18.64    NaN    NaN    NaN  
2820   1.19   6.54  13.76    NaN    NaN    NaN  
2962   1.16   6.95  15.49    NaN    NaN    NaN  
3066   1.20   6.16  14.76    NaN    NaN    NaN

**Mecze gdzie mocnymi faworytami byli gospodarze (kursy mniejsze niż 1.30) a przegrwywali mecz**

In [155]:
df[(df["PSCH"] < 1.3) & (df["Res"] == "A")]

Country       League     Season       Date      Time            Home  \
1301  Poland  Ekstraklasa  2016/2017 2017-02-19  17:00:00           Legia   
3276  Poland  Ekstraklasa  2023/2024 2023-10-29  14:00:00           Legia   
3300  Poland  Ekstraklasa  2023/2024 2023-11-25  16:30:00  Pogon Szczecin   

             Away  HG  AG Res  PSCH  PSCD   PSCA  MaxCH  MaxCD  MaxCA  AvgCH  \
1301         Ruch   1   3   A  1.26  6.28  13.56   1.27   6.30  14.36   1.24   
3276  Stal Mielec   1   3   A  1.28  5.96  11.47   1.39   6.02  11.47   1.28   
3300  Stal Mielec   2   3   A  1.28  6.06  10.56   1.32   6.20  11.00   1.27   

      AvgCD  AvgCA  BFECH  BFECD  BFECA  
1301   5.71  11.36    NaN    NaN    NaN  
3276   5.43   9.52    NaN    NaN    NaN  
3300   5.58   9.63    NaN    NaN    NaN

**Mecze z udziałem Legii, Lecha i Cracovii**

In [156]:
teams = ["Legia", "Lech Poznan", "Cracovia"]
df[(df["Home"].isin(teams)) | (df["Away"].isin(teams))]

Country       League     Season       Date      Time              Home  \
4     Poland  Ekstraklasa  2012/2013 2012-08-18  17:00:00       Lech Poznan   
6     Poland  Ekstraklasa  2012/2013 2012-08-19  16:00:00             Legia   
12    Poland  Ekstraklasa  2012/2013 2012-08-25  17:00:00  Polonia Warszawa   
14    Poland  Ekstraklasa  2012/2013 2012-08-26  16:00:00     GKS Belchatow   
21    Poland  Ekstraklasa  2012/2013 2012-09-02  13:30:00       Lech Poznan   
...      ...          ...        ...        ...       ...               ...   
3764  Poland  Ekstraklasa  2024/2025 2025-05-18  13:45:00          Cracovia   
3765  Poland  Ekstraklasa  2024/2025 2025-05-18  16:30:00      GKS Katowice   
3769  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00       Lech Poznan   
3771  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00             Legia   
3775  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00          Zaglebie   

               Away  HG  AG Res  PSCH  PSCD  PSCA  MaxCH  MaxCD  MaxCA  AvgCH  \
4              Ruch   4   0   H  2.12  3.29  3.89   2.12   3.30   4.00   2.05   
6     Korona Kielce   4   0   H  1.63  3.67  7.21   1.70   3.71   7.22   1.63   
12      Lech Poznan   1   2   A  2.97  3.20  2.62   3.10   3.25   2.74   2.83   
14            Legia   0   2   A  4.78  3.38  1.93   4.78   3.45   2.00   3.92   
21        Gornik Z.   0   0   D  1.65  3.88  5.98   1.65   4.09   6.75   1.58   
...             ...  ..  ..  ..   ...   ...   ...    ...    ...    ...    ...   
3764          Legia   3   1   H  2.79  3.62  2.41   2.94   3.65   2.48   2.75   
3765    Lech Poznan   2   2   D  5.27  4.28  1.58   5.54   4.35   1.63   5.24   
3769  Piast Gliwice   1   0   H  1.33  5.27  8.47   1.38   5.60   9.00   1.34   
3771    Stal Mielec   2   2   D  1.35  5.47  7.60   1.44   5.50   7.70   1.36   
3775       Cracovia   1   2   A  2.77  3.45  2.51   2.77   3.55   2.56   2.67   

      AvgCD  AvgCA  BFECH  BFECD  BFECA  
4      3.16   3.61    NaN    NaN    NaN  
6      3.49   5.66    NaN    NaN    NaN  
12     3.06   2.48    NaN    NaN    NaN  
14     3.28   1.93    NaN    NaN    NaN  
21     3.58   5.99    NaN    NaN    NaN  
...     ...    ...    ...    ...    ...  
3764   3.53   2.39   2.94   3.70   2.52  
3765   4.10   1.59   5.90   4.30   1.66  
3769   5.18   8.20   1.36   5.90  10.00  
3771   5.29   7.12   1.40   5.60   8.80  
3775   3.45   2.49   2.86   3.65   2.60  

[1248 rows x 22 columns]

**Mecze najbardziej bramkoszczelne w przypadku gospodarzy i gośći minimum po 4 bramki dla zespołu u siebie i na wyjeździe**

In [157]:
specific_match=df[(df['HG'] >= 4) & (df['AG'] >= 4)]
specific_match

Country       League     Season       Date      Time           Home  \
206   Poland  Ekstraklasa  2012/2013 2013-05-12  16:00:00  Lechia Gdansk   
520   Poland  Ekstraklasa  2013/2014 2014-05-27  19:30:00    Jagiellonia   
2011  Poland  Ekstraklasa  2018/2019 2019-05-18  14:30:00          Wisla   
2084  Poland  Ekstraklasa  2019/2020 2019-09-21  19:00:00  Slask Wroclaw   
2218  Poland  Ekstraklasa  2019/2020 2020-03-07  16:30:00       Zaglebie   
3310  Poland  Ekstraklasa  2023/2024 2023-12-03  11:30:00       Cracovia   
3731  Poland  Ekstraklasa  2024/2025 2025-04-25  17:00:00        Puszcza   

                Away  HG  AG Res  PSCH  PSCD  PSCA  MaxCH  MaxCD  MaxCA  \
206             Ruch   4   4   D  2.38  2.94  3.77   2.38   3.25   3.77   
520    Korona Kielce   4   4   D  2.73  3.25  2.81   2.73   3.30   2.92   
2011         Legnica   4   5   A  2.54  4.05  2.56   2.58   4.05   2.76   
2084        Zaglebie   4   4   D  2.10  3.43  3.74   2.13   3.60   3.75   
2218   Lechia Gdansk   4   4   D  1.73  4.09  4.56   1.93   4.09   4.56   
3310    Ruch Chorzow   4   4   D  1.67  3.81  5.67   1.76   4.10   6.40   
3731  Pogon Szczecin   4   5   A  4.77  4.11  1.69   4.77   4.25   1.71   

      AvgCH  AvgCD  AvgCA  BFECH  BFECD  BFECA  
206    2.26   3.05   3.23    NaN    NaN    NaN  
520    2.50   3.17   2.77    NaN    NaN    NaN  
2011   2.46   3.62   2.59    NaN    NaN    NaN  
2084   2.04   3.35   3.55    NaN    NaN    NaN  
2218   1.80   3.64   4.18    NaN    NaN    NaN  
3310   1.62   3.74   5.41    NaN    NaN    NaN  
3731   4.58   4.02   1.67    5.1    4.2   1.73

**Najwięcej strzelonych bramek w meczu**

In [158]:
df['TotalGoals'] = df['HG'] + df['AG']
max_goals = df['TotalGoals'].max()
most_goals_matches = df[df['TotalGoals'] == max_goals]
print(f"Najwięcej bramek w jednym meczu: {max_goals}")
most_goals_matches

Najwięcej bramek w jednym meczu: 9


Country       League     Season       Date      Time           Home  \
795   Poland  Ekstraklasa  2014/2015 2015-05-19  19:30:00  Piast Gliwice   
2011  Poland  Ekstraklasa  2018/2019 2019-05-18  14:30:00          Wisla   
3731  Poland  Ekstraklasa  2024/2025 2025-04-25  17:00:00        Puszcza   
3742  Poland  Ekstraklasa  2024/2025 2025-05-03  19:15:00    Lech Poznan   

                Away  HG  AG Res  PSCH  PSCD  PSCA  MaxCH  MaxCD  MaxCA  \
795    GKS Belchatow   6   3   H  1.80  3.69  5.02   1.81   4.10   5.08   
2011         Legnica   4   5   A  2.54  4.05  2.56   2.58   4.05   2.76   
3731  Pogon Szczecin   4   5   A  4.77  4.11  1.69   4.77   4.25   1.71   
3742         Puszcza   8   1   H  1.31  5.64  9.54   1.32   6.25  10.00   

      AvgCH  AvgCD  AvgCA  BFECH  BFECD  BFECA  TotalGoals  
795    1.75   3.52   4.64    NaN    NaN    NaN           9  
2011   2.46   3.62   2.59    NaN    NaN    NaN           9  
3731   4.58   4.02   1.67   5.10    4.2   1.73           9  
3742   1.28   5.55   9.18   1.32    6.2  10.00           9

**Najbardziej bramkoszczelna drużyna**

In [159]:
home_goals = df.groupby('Home')['HG'].sum()
away_goals = df.groupby('Away')['AG'].sum()
total_scored = home_goals.add(away_goals, fill_value=0)
print ("Najwięcej strzelonych bramek w meczu:")
print(total_scored.sort_values(ascending=False).head(10))

Najwięcej strzelonych bramek w meczu:
Home
Legia             780
Lech Poznan       729
Jagiellonia       676
Pogon Szczecin    632
Slask Wroclaw     584
Piast Gliwice     576
Cracovia          561
Lechia Gdansk     555
Zaglebie          531
Wisla             488
dtype: int64


**Która drużyna zdobyła najwięcej punktów jako gospodarz w badanym okresie?**

In [160]:
df['HomePoints'] = df['Res'].map({'H': 3, 'D': 1, 'A': 0})
home_points = df.groupby('Home')['HomePoints'].sum().sort_values(ascending=False)
print("Najwięcej punktów zdobytych jako gospodarz:")
print(home_points.head(10))

Najwięcej punktów zdobytych jako gospodarz:
Home
Legia             480
Lech Poznan       454
Pogon Szczecin    389
Jagiellonia       389
Piast Gliwice     380
Slask Wroclaw     372
Lechia Gdansk     344
Cracovia          337
Korona Kielce     301
Zaglebie          299
Name: HomePoints, dtype: int64


**Średni kurs na wygraną gospodarzy**

In [189]:
avg_course_to_hometeams = df['PSCH'].mean().__round__(4)
avg_course_to_hometeams


np.float64(2.5073)

**Średni kurs na remis**

In [188]:
avg_course_to_draws = df['PSCD'].mean().__round__(4)
avg_course_to_draws

np.float64(3.6161)

**Średni kurs na wygraną gości**

In [192]:
avg_course_to_awayteams = df['PSCA'].mean().__round__(4)
avg_course_to_awayteams

np.float64(3.8896)

**Średnia ilość bramek na sezon**

In [ ]:
df.groupby('Season')[['HG', 'AG']].mean()

HG        AG
Season                       
2012/2013  1.341667  1.150000
2013/2014  1.540541  1.114865
2014/2015  1.493243  1.175676
2015/2016  1.456081  1.172297
2016/2017  1.577703  1.165541
2017/2018  1.479730  1.158784
2018/2019  1.500000  1.280405
2019/2020  1.516892  1.074324
2020/2021  1.304167  1.150000
2021/2022  1.500000  1.173203
2022/2023  1.372549  1.104575
2023/2024  1.545752  1.143791
2024/2025  1.542484  1.212418

**Suma wszystkich bramek na sezon**

In [226]:
df.groupby('Season')[['HG', 'AG']].sum().assign(TotalGoals=lambda x: x['HG'] + x['AG'])

HG   AG  TotalGoals
Season                         
2012/2013  322  276         598
2013/2014  456  330         786
2014/2015  442  348         790
2015/2016  431  347         778
2016/2017  467  345         812
2017/2018  438  343         781
2018/2019  444  379         823
2019/2020  449  318         767
2020/2021  313  276         589
2021/2022  459  359         818
2022/2023  420  338         758
2023/2024  473  350         823
2024/2025  472  371         843

In [244]:
df.groupby('Season')[['HG', 'AG']].sum().sum(axis=1).to_frame('TotalGoals')

TotalGoals
Season               
2012/2013         598
2013/2014         786
2014/2015         790
2015/2016         778
2016/2017         812
2017/2018         781
2018/2019         823
2019/2020         767
2020/2021         589
2021/2022         818
2022/2023         758
2023/2024         823
2024/2025         843

**Średnie kursy w każdym kolejnym seoznie na wygraną gospodarzy, wygraną gośći i remis**

In [261]:
df.groupby('Season')[['PSCH', 'PSCA','PSCD']].mean().round(2)

PSCH  PSCA  PSCD
Season                     
2012/2013  2.47  4.14  3.49
2013/2014  2.50  4.04  3.53
2014/2015  2.46  4.00  3.58
2015/2016  2.50  3.92  3.58
2016/2017  2.50  4.06  3.64
2017/2018  2.41  3.87  3.60
2018/2019  2.38  3.63  3.59
2019/2020  2.42  3.67  3.62
2020/2021  2.69  3.55  3.65
2021/2022  2.65  4.07  3.70
2022/2023  2.60  3.85  3.66
2023/2024  2.56  3.95  3.68
2024/2025  2.47  3.78  3.67

**Najniższe kursy w poszczegołnych sezonach**

In [273]:
df.groupby('Season')[['PSCH','PSCA','PSCD']].max()

PSCH   PSCA  PSCD
Season                       
2012/2013   7.35  17.32  6.54
2013/2014   7.91  12.02  5.68
2014/2015   7.10  13.80  6.75
2015/2016   6.18  15.53  6.20
2016/2017   8.44  17.87  7.21
2017/2018   6.55  11.44  5.41
2018/2019   6.32  11.35  6.61
2019/2020   6.65  15.17  6.84
2020/2021   8.57  12.17  6.13
2021/2022  12.81  23.43  7.79
2022/2023   8.81  18.67  7.87
2023/2024   8.05  11.53  6.06
2024/2025   8.22  11.81  5.65

**Najniższe kursy w poszczegołnych sezonach**

In [274]:
df.groupby('Season')[['PSCH','PSCA','PSCD']].min()

PSCH  PSCA  PSCD
Season                     
2012/2013  1.23  1.56  2.90
2013/2014  1.31  1.49  2.91
2014/2015  1.24  1.56  3.01
2015/2016  1.25  1.59  2.92
2016/2017  1.20  1.46  2.99
2017/2018  1.32  1.62  2.88
2018/2019  1.24  1.54  2.94
2019/2020  1.22  1.47  2.96
2020/2021  1.28  1.43  2.91
2021/2022  1.16  1.26  2.80
2022/2023  1.16  1.44  2.79
2023/2024  1.28  1.43  2.78
2024/2025  1.26  1.41  2.70

In [275]:
df

Country       League     Season       Date      Time            Home  \
0     Poland  Ekstraklasa  2012/2013 2012-08-17  17:00:00  Pogon Szczecin   
1     Poland  Ekstraklasa  2012/2013 2012-08-17  19:30:00   Piast Gliwice   
2     Poland  Ekstraklasa  2012/2013 2012-08-18  12:30:00     Jagiellonia   
3     Poland  Ekstraklasa  2012/2013 2012-08-18  14:45:00     Widzew Lodz   
4     Poland  Ekstraklasa  2012/2013 2012-08-18  17:00:00     Lech Poznan   
...      ...          ...        ...        ...       ...             ...   
3771  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00           Legia   
3772  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00         Puszcza   
3773  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00  Radomiak Radom   
3774  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00           Rakow   
3775  Poland  Ekstraklasa  2024/2025 2025-05-24  16:30:00        Zaglebie   

               Away  HG  AG Res  PSCH  PSCD   PSCA  MaxCH  MaxCD  MaxCA  \
0          Zaglebie   4   0   H  3.23  3.21   2.47   3.30   3.30   2.50   
1         Gornik Z.   1   2   A  2.72  3.12   2.98   2.78   3.25   3.30   
2      Podbeskidzie   2   1   H  1.74  3.60   5.63   1.85   3.60   5.63   
3     Slask Wroclaw   2   1   H  2.70  3.11   2.95   2.80   3.20   2.95   
4              Ruch   4   0   H  2.12  3.29   3.89   2.12   3.30   4.00   
...             ...  ..  ..  ..   ...   ...    ...    ...    ...    ...   
3771    Stal Mielec   2   2   D  1.35  5.47   7.60   1.44   5.50   7.70   
3772  Slask Wroclaw   1   1   D  3.25  3.93   2.05   3.30   4.00   2.05   
3773   Motor Lublin   2   3   A  1.93  3.64   3.86   1.99   3.75   4.05   
3774    Widzew Lodz   2   1   H  1.26  5.64  11.81   1.29   5.80  11.81   
3775       Cracovia   1   2   A  2.77  3.45   2.51   2.77   3.55   2.56   

      AvgCH  AvgCD  AvgCA  BFECH  BFECD  BFECA  TotalGoals  HomePoints  all  
0      3.01   3.11   2.34    NaN    NaN    NaN           4           3    4  
1      2.56   3.05   2.78    NaN    NaN    NaN           3           0    3  
2      1.74   3.42   4.72    NaN    NaN    NaN           3           3    3  
3      2.66   3.03   2.66    NaN    NaN    NaN           3           3    3  
4      2.05   3.16   3.61    NaN    NaN    NaN           4           3    4  
...     ...    ...    ...    ...    ...    ...         ...         ...  ...  
3771   1.36   5.29   7.12   1.40   5.60   8.80           4           1    4  
3772   3.24   3.88   2.01   3.55   4.10   2.08           2           1    2  
3773   1.94   3.54   3.73   2.02   3.80   3.90           5           0    5  
3774   1.27   5.47  10.29   1.30   6.20  13.50           3           3    3  
3775   2.67   3.45   2.49   2.86   3.65   2.60           3           0    3  

[3776 rows x 25 columns]